lib/active_model/serializer/fieldset.rb

```diff
@@ -18,14 +18,23 @@ def fields_for(type)
       attr_reader :raw_fields
 
       private
-
       def parsed_fields
+        # TODO: this is not very flexible.
+        # - we should probably support symbol, array, hash, etc
+        #
+        # JSONAPI::IncludeDirective supports these, but also changes the resulting structure.
+        # is it worth it to duplicate some of that functionality?
+        # or should IncludeDirective be modified to only allow one layer of options?
```

Comment:
Related https://github.com/rails-api/active_model_serializers/pull/1857

While working on that I've discovered an architectural problem that will require some careful unraveling

In brief

serializer.attributes(fields) builds a hash from calling value (aka read_attribute_for_serialization ) on the Attribute objects with matching names

serializer.associations(include_directive) iterates over all the Reflection objects and rejects excluded and yields an Association object built jit by the Reflection

The Association is used by calling association.serializer.serializable_hash

As currently implemented, the reflection is evaluated (method is called) when the Association is built.  This prevents lazy evaluation

I have a hack in that pr to delay evaluation until the serializer asks for it, which works when the serializer calls serializable_hash

However, the JSONAPI adapter iterates the attributes and associations directly and passes the association into a new Relationship which promptly evaluates the association before determining whether to include it.  Complicating this is that there is an assumption in the code that we mist evaluate the association so that we can infer the serializer class so that we can instantiate it, but it is already too late.

The fundamental way that a serializer is told which fields and associations to evaluate is inconsistent and far to eager.  The already-evaluated-association object permeates the code.... Should be much simpler

Also would be good to make Fields and similar apis consistent